urld2fb1d2b05_thnaeff_Util_tgz-pJ8-ch_thn_util_tests_TreeTestJ8/src/ch/thn/util/collections/MultiKeyHashMap.java

```diff
@@ -0,0 +1,306 @@
+			}
+		}
+		
+		mapping.put(newKey, mapTo);
+		
+		if (!mappingReverse.containsKey(mapTo)) {
+			mappingReverse.put(mapTo, new HashSet<K>(1));
+		}
+		
+		mappingReverse.get(mapTo).add(newKey);
+	}
+	
+	/**
+	 * Returns the key to which <code>newKey</code> points to.<br>
+	 * The returned key is the one which has been used with <code>mapTo</code> 
+	 * in {@link #addKeyMapping(Object, Object)}
+	 * 
+	 * @param newKey
+	 * @return
+	 */
+	public K getMappedTo(K newKey) {
+		return mapping.get(newKey);
+	}
+	
+	/**
+	 * Returns the farthest key to which the given key points to.<br>
+	 * The returned key is the "original" key which is the key that is actually 
+	 * used in the hash map to reference to the value.<br>
+	 * Since mappings can be mapped to other mappings, this method travels 
+	 * through all the mappings until a mapped key matches one in the hash map
+	 * 
+	 * @param key
+	 * @return
+	 */
+	public K getMapping(K key) {
+		return getKeyK(key);
+	}
+	
+	/**
+	 * Returns all the keys which point to the key <code>mapTo</code>
+	 * 
+	 * @param mapTo
+	 * @return
+	 */
+	public HashSet<K> getReverseMapping(K mapTo) {
+		return mappingReverse.get(mapTo);
+	}
+	
+	/**
+	 * Returns the "original" key which is used in the hash map
+	 * 
+	 * @param key
+	 * @return
+	 */
+	private Object getKeyObject(Object key) {
+		Object returnKey = key;
+		
+		while (!super.containsKey(returnKey)) {
+			if (mapping.containsKey(returnKey)) {
+				returnKey = mapping.get(returnKey);
+			} else {
+				break;
+			}
+		}
+		
+		return returnKey;
+	}
+	
+	/**
+	 * Returns the "original" key which is used in the hash map
+	 * 
+	 * @param key
+	 * @return
+	 */
+	private K getKeyK(K key) {
+		K returnKey = key;
+		
+		while (!super.containsKey(returnKey)) {
+			if (mapping.containsKey(returnKey)) {
+				returnKey = mapping.get(returnKey);
+			} else {
+				break;
+			}
+		}
+		
+		return returnKey;
+	}
+	
+	/**
+	 * Copies the mappings of the given {@link MultiKeyHashMap} to this {@link MultiKeyHashMap}
+	 * 
+	 * @param m
+	 */
+	@SuppressWarnings({ "rawtypes", "unchecked" })	//Only the mapping fields are accessed, thus the type of the map does not matter
+	public void copyMapping(MultiKeyHashMap m) {
+		mapping.putAll(m.mapping);
+		mappingReverse.putAll(m.mappingReverse);
+	}
+	
+	@Override
+	public V get(Object key) {
+		return super.get(getKeyObject(key));
+	}
+	
+	@Override
+	public boolean containsKey(Object key) {
+		return super.containsKey(getKeyObject(key));
+	}
+	
+	@Override
+	public V put(K key, V value) {
+		return super.put(getKeyK(key), value);
+	}
+	
+	@Override
+	public void putAll(Map<? extends K, ? extends V> m) {
+		Map<K, V> m2 = new HashMap<K, V>(m.size());
+		
+		//Create a new map and replace all keys with their mapping if there is 
+		//a mapping available
+		for (K key : m.keySet()) {
+			if (mapping.containsKey(key)) {
+				m2.put(mapping.get(key), m.get(key));
+			} else {
+				m2.put(key, m.get(key));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.